Sources/SourceKitLSP/Clang/ClangLanguageServer.swift

```diff
@@ -137,6 +140,7 @@ final class ClangLanguageServerShim: LanguageServer, ToolchainLanguageServer {
     process.standardInput = usToClangd
     process.terminationHandler = { [weak self] process in
       log("clangd exited: \(process.terminationReason) \(process.terminationStatus)")
+      self?.clangPid = nil
```

Comment:
I think this mutation needs to move to `self.queue.async`.  Maybe something like this (not tested)?

```swift
connectionToClangd.close()
self?.queue.async { [self] in
  self.clangPid = nil
  if process.terminationStatus != 0 {
    self.state = .connectionInterrupted
    self.restartClangd()
  }
}
```

And since I'm looking at it, _crash ought to be `queue` as well.  I think async is fine, since the test is going to need to wait on it coming back anyway.  If we ever need to make a crash function that will interrupt an ongoing operation we could lock the pid variable separately from queue, but it doesn't seem like it matters right now.

Reply:
I can't really imagine a race here, but sounds good to better be safe and make the crash be polite and patiently wait in the queue to kill the process 😉